packages/core/src/browser/shell/application-shell.ts

```diff
@@ -1076,25 +1077,27 @@ export class ApplicationShell extends Widget {
         if (!current) {
             return undefined;
         }
-        await Promise.all([
+        return Promise.all([
             this.waitForActivation(current.id),
             waitForRevealed(current),
             this.pendingUpdates
-        ]);
-        return current;
+        ]).then(() => current, () => undefined);
     }
 
     waitForActivation(id: string): Promise<void> {
         if (this.activeWidget && this.activeWidget.id === id) {
             return Promise.resolve();
         }
-        return new Promise(resolve => {
-            const toDispose = this.onDidChangeActiveWidget(() => {
-                if (this.activeWidget && this.activeWidget.id === id) {
-                    toDispose.dispose();
-                    resolve();
-                }
-            });
+        const activation = new Deferred();
+        const success = this.onDidChangeActiveWidget(() => {
+            if (this.activeWidget && this.activeWidget.id === id) {
+                activation.resolve();
+            }
+        });
+        const failure = setTimeout(() => activation.reject(new Error(`Widget with id '${id}' failed to activate.`)), this.activationTimeout + 250);
+        return activation.promise.finally(() => {
+            success.dispose();
+            clearTimeout(failure);
         });
```

Comment:
```suggestion
        return new Promise((resolve, reject) => {
            let listener, timeout;
            listener = this.onDidChangeActiveWidget(() => {
                if (this.activeWidget && this.activeWidget.id === id) {
                    clearTimeout(timeout);
                    listener.dispose();
                    resolve();
                }
            });
            timeout = setTimeout(() => {
                listener.dispose();
                reject(new Error(`Widget with id '${id}' failed to activate.`))
            }, this.activationTimeout + 250);
        });
```

I rarely see the use for `Deferred` objects...

Reply:
I don't really see any advantage to rewriting it like this. The main advantages of the `Deferred` in this case to my mind are that (mainly) you can refer to the things you need to dispose of in a `finally` rather than having to write the cleanup as part of the resolution or rejection and (much less importantly), it saves one level of indentation and mental effort tracking what's available in what scope.